src/include/detail/graph/nn-graph.h

```diff
@@ -0,0 +1,231 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ *
+ */
+
+#ifndef TILEDB_GRAPH_H
+#define TILEDB_GRAPH_H
+
+#include <vector>
+#include "scoring.h"
+#include "utils/fixed_min_heap.h"
+
+namespace detail::graph {
+
+template <class SC, std::integral ID>
+class nn_graph {
+ public:
+  using id_type = ID;
+  using score_type = SC;
+
+ private:
+  size_t num_vertices_{0};
+  size_t k_nn_{0};
+
+  std::vector<fixed_min_pair_heap<score_type, id_type>> out_edges_;
+  std::vector<std::set<ID>> in_edges_;
+  std::vector<fixed_min_pair_heap<score_type, id_type>> update_edges_;
+
+ public:
+  nn_graph() = default;
+
+  nn_graph(size_t num_vertices, size_t k_nn)
+      : num_vertices_{num_vertices}
+      , k_nn_{k_nn}
+      , out_edges_(num_vertices, fixed_min_pair_heap<score_type, id_type>(k_nn))
+      , in_edges_(num_vertices)
+      , update_edges_(
+            num_vertices, fixed_min_pair_heap<score_type, id_type>(k_nn)) {
+  }
+
+  /**
+   * Attempt to add edge to out_edge neighbor list.  Note that we can't add a
+   * corresponding edge to the in_edges_ list because we don't know which edge
+   * may have been displaced if the insert was successful.
+   * @param src
+   * @param dst
+   * @param score
+   */
+  auto add_edge(id_type src, id_type dst, score_type score) {
+    // return out_edges_[src]. template insert<typename
+    // std::remove_cvref_t<decltype(out_edges_[src])>::unique_id>(score, dst);
+    return out_edges_[src].template insert<unique_id>(score, dst);
+  }
+
+  /**
+   * Attempt to add edge to update_edge neighbor list.  Note that we can't add a
+   * corresponding edge to the in_edges_ list because we don't know which edge
+   * may have been displaced if the insert was successful.
+   * @param src
+   * @param dst
+   * @param score
+   */
+  auto add_update_edge(id_type src, id_type dst, score_type score) {
+    return update_edges_[src]
+        .template insert<
+            typename std::remove_cvref_t<decltype(out_edges_[src])>::unique_id>(
+            score, dst);
+  }
+
+  auto copy_to_update_edges() {
+    for (size_t i = 0; i < num_vertices_; ++i) {
+      for (auto&& [s, e] : out_edges_[i]) {
+        update_edges_[i].insert(s, e);
+      }
+    }
+  }
+
+  auto swap_update_edges(id_type i) {
+    out_edges_[i].swap(update_edges_[i]);
+  }
+
+  auto swap_all_update_edges() {
+    for (id_type i = 0; i < num_vertices_; ++i) {
+      swap_update_edges(i);
+    }
+  }
+
+  auto sort_edges(id_type i) {
+  }
```

Comment:
Remove? Or if it's intentionally a no-op, let's add a comment saying that, please.